### DISCUSSION TITLE
set_parameter pmin and pmax and ml.plots.results

### DISCUSSION THREAD

**Discussion by IsaDio** [pastas/pastas#324] [Q&A] [answered]
Hi,

Just realised that my previous question might have been a bit vague. Sorry about this.

I have been trying to calibrate time series where the trend in heads has been going up due to climate.
I only have 5 years of data at most bores but 1 bore has an additional 10 years of data. All bores show similiar trends and heads.
I tested many models (including nonlinear recharge models) but results were non conclusive, which is why I asked my previous question. I have been doing some testing since and I think I am in the right direction.

At the moment the best results are obtained when I start the calibration in the 1980s to get long term climate variation and if I set a range (pmin and pmax) for "constant_d" using ml.set_parameter. This allows to set the initial conditions approximately (unknown but we have an idea). This has giving me much better results.

The only issue at the moment is if I set tmin=1980 for ml.plots.results I get different results in the time series plot. Parameters are the same but the simulated time series is shifted up (see second figure below). I don't get that issue if I don't set pmin and pmax.
Is it something I am doing wrong?

Thank you for your help.

![MB16WAW0005_solve_1](https://user-images.githubusercontent.com/91039837/135056497-e8efc387-212f-49f2-b6d0-0c12cf66d2e7.png)
![MB16WAW0005_solve_2](https://user-images.githubusercontent.com/91039837/135056475-daf77fbf-d90a-4d12-bfd4-0bf706ff4b28.png)

**Comment by rubencalje** [comment, marked as answer]
Hi @IsaDio,

In theory non-stationary models can be evaluated using Pastas, but the response curve (especially the tail) will be uncertain. Also your stress-series (precipitation and evaporation) should be long enough (into the past). What you show, with a different simulation for a different tmin, seems like a bug. This however can be explained by the warmup you use, which has a default value of 3650 days (10 years). In your case, this is too short, which causes a different simulation when you start in 2016 (minus the warmup: 2006) compared to 1980 (minus the warmup: 1970). You can try changing the warmup period, so it reflects your entire period with precipitation-data (40 years?):

ml.solve(noise=False, warmup=365*40)

Another comment: the evaporation-factor is almost 0, which means evaporation has no effect. Is this intentional?

Regards,

Ruben

**Reply by IsaDio** [reply]
Hi @rubencalje,

Thanks for the quick reply.
Yes that makes sense for the long stress-series, as we want to capture that historical change to get good initial conditions (from when head data are available).
When I solve the model I use the same tmin, ml.solve(noise=False, tmin='1980'). My understanding is the warmup is always adjusted as 10 years before the tmin, and I get the above error. Shouldn't that do the same as changing the warmup period?
I tried modifying the warmup period instead of tmin with what you suggested and I still get the same issue with StressModel2, but not with RechargeModel. When I look at the outputs it looks like the stress-series are extended twice 36 years. Is it possible the warmup period also affects ml.plots.results? 
Sorry if this is something I am doing wrong but I can't seem to understand what it's doing.

![Capture](https://user-images.githubusercontent.com/91039837/135095376-dc890cac-51ab-4f69-a056-50d35275aca4.PNG)

Yes I would expect evaporation to have little effect. The water table is deep (25m) and rainfall (from significant rainfall events and cyclones) is the main driver to recharge. I included it as it is required for the nonlinear recharge models.

Thanks again for you help. Really appreciated.

Isabelle

**Reply by rubencalje** [reply]
`StressModel2 `is the predecessor of `RechargeModel`, which does the same and more. We mostly keep `StressModel2 `for backwards-compatibility, so I think you should use `RechargeModel`.

The warmup-period should ideally be at least as long as your response time. The warmup period is added before tmin, and the simulation in the warmup period is normally not used (unless `return_warmup `is True in some methods): it is not plotted and measurements in the warmup period are ignored. In that way it is different than just changing tmin.

Whenever you run `ml.solve()`, `ml.plots.results() `or `ml.simulate()`, the model reforms the stress-input-series, to make sure it is of the right length and frequency. When tmin minus the warmup period extends before the start of your precipitation-series, it needs to fill in some value at the start of the stress. For a precipitation-series it takes the mean value of the data between tmin-warmup and tmax for this period (I believe, this setting can be viewed using `ps.rcParams["timeseries"]`). This causes the fact that a different fill_before-value is used when you specify tmin='2016' compared to tmin='1980'. Because of the high response time, this has a large effect on the entire simulation. I hope this somehow explains what the model is doing, and what causes the strange results.

Regards,

Ruben

**Reply by IsaDio** [reply]
Hi @rubencalje, thank you for the explanation. 
This is what I figured out with ml.solve() and ml.plots.results().
What I can't figure out is why using tmin='1980' for both ml.solve() and ml.plots.results() doesn't give the same results (I have rainfall data from 01/1970). The optimal parameters are the same but the simulation is different (shift like the figure in my original question, which used the same tmin for ml.solve() and ml.plots.results()).  Looking at the fit report, it looks like tmin is always set to the first wl data point (04/2016) even if you fix it in ml.solve()?
I can't get things to consistently work, its often all over the place depending on tmin and warmup period (using sensible numbers). Which makes it hard to understand what the Recharge Model is actually doing.

Sorry if it is something I am doing wrong but still can't make sense of it.

**Reply by martinvonk** [reply]
Hi @IsaDio,

What you show, with a different simulation (in ml.solve and ml.plots.results) for a different tmin, seems like a bug. You can create a similar plot as _ml.plots.results_ by yourself using `ml.solve(tmin='1980')` and plotting `ml.simulate(tmin='1980')` (simulated time series) & `ml.observations()` (observed time series). Hopefully you don't have the vertical shift then.

**Reply by IsaDio** [reply]
Thanks @martinvonk. 
I think I understand what is happening. Maybe I should have figured it out earlier... Setting tmin in ml.solve(tmin='1980') doesn't do anything. It's always using 2016-04-15 (first head observation). I get the same calibration results if I set tmin='1980' and if I don't set it in ml.solve. So it looks like I can only change the warmup period if I want to start the simulation earlier. I can use ml.simulate(return_warmup=True) to get the full time series. That works fine :)

Sorry I forgot to say because setting tmin in ml.solve doesn't work, I still end up with the same shift using ml.simulate(tmin='1980') than with ml.plots.results(tmin='1980').
